3p/environment.js

```diff
@@ -0,0 +1,241 @@
+      if (node.tagName != 'IFRAME') {
+        continue;
+      }
+      const src = node.getAttribute('src');
+      const srcdoc = node.getAttribute('srcdoc');
+      if (src == null || /^(about:|javascript:)/i.test(src.trim()) ||
+          srcdoc) {
+        if (node.contentWindow) {
+          instrumentIframeWindow(node, win, node.contentWindow, depth);
+          node.addEventListener('load', () => {
+            instrumentIframeWindow(node, win, node.contentWindow, depth);
+          });
+        }
+        else if (srcdoc) {
+          instrumentSrcdoc(parent, node, depth);
+        }
+      } else {
+        console.info('Skipping cross origin iframe ',src);
+      }
+    } catch (e) {
+      console.error(e.message, e.stack);
+    }
+  }
+}
+
+/**
+ * Installs a mutation observer in a window to look for iframes.
+ * @param {!Element} node
+ * @param {!Window} parent
+ * @param {!Window} win
+ * @param {number} depth
+ */
+function instrumentIframeWindow(node, parent, win, depth) {
+  if (win.ampSeen) {
+    return;
+  }
+  const doc = win.document;
+  console.info('iframe candidate', node, doc.body);
+  instrumentDocWrite(parent, win, depth)
+  if (doc.body && doc.body.childNodes.length) {
+    manageWin(win, depth + 1);
+  }
+}
+
+/**
+ * Installs a mutation observer in a window to look for iframes.
+ * @param {!Window} win
+ * @param {number} depth
+ */
+function installObserver(win, depth) {
+  const observer = new MutationObserver(function(mutations) {
+    for (let i = 0; i < mutations.length; i++) {
+      maybeInstrumentsNodes(win, mutations[i].addedNodes, depth);
+    }
+  });
+  observer.observe(win.document.documentElement, {
+    subtree: true,
+    childList: true,
+  });
+}
+
+/**
+ * Replace timers with variants that can be throttled.
+ * @param {!Window} win
+ */
+function instrumentEntryPoints(win) {
+  // Change setTimeout to respect a minimum timeout.
+  const setTimeout = win.setTimeout;
+  win.setTimeout = function(fn, time) {
+    time = minTime(time);
+    return setTimeout(fn, time);
+  };
+  // Implement setInterval in terms of setTimeout to make
+  // it respect the same rules
+  const setInterval = win.setInterval;
+  const intervals = {};
+  let intervalId = 0;
+  win.setInterval = function(fn, time) {
+    const id = intervalId++;
+    function next() {
+      intervals[id] = win.setTimeout(function() {
+        next();
+        return fn.apply(this, arguments);
+      }, time);
+    }
+    next();
+    return id;
+  };
+  win.clearInterval = function(id) {
+    win.clearTimeout(intervals[id]);
+    delete intervals[id];
+  };
+  // Throttle requestAnimationFrame.
+  const requestAnimationFrame = win.requestAnimationFrame ||
+      win.webkitRequestAnimationFrame;
+  win.requestAnimationFrame = function() {
+    //console.info('requestAnimationFrame', arguments);
+    if (!docState.inViewport) {
+      // Do not animate when not in viewport.
+      return requestAnimationFrame.call(this, emptyFunction);
```

Comment:
I thought it just slowed down the rate? Here, you're actively breaking the recursive chain so that the `fn` can't call `requestAnimationFrame(fn)` to reschedule itself on the next frame.